crates/valence_nbt/src/snbt.rs

```diff
@@ -0,0 +1,685 @@
+        match self.peek()? {
+            '{' => self.parse_compound().map(|c| c.into()),
+            '[' => self.parse_list_like(),
+            '"' | '\'' => self.read_quoted_string().map(|s| s.into()),
+            _ => self.parse_primitive(),
+        }
+    }
+
+    pub fn read(&mut self) -> Result<Value> {
+        let value = self.parse_element()?;
+        self.skip_whitespace();
+        if self.peek().is_ok() {
+            return Err(self.make_error(SNBTErrorType::TrailingData));
+        }
+        Ok(value)
+    }
+
+    /// Get the number of bytes readed.
+    /// It's useful when you want to read a SNBT string from an command argument since there may be trailing data.
+    pub fn bytes_readed(&self) -> usize {
+        self.index
+    }
+
+    /// Parse a string in SNBT format into a `Value`.
+    /// Assert that the string has no trailing data.
+    /// SNBT is quite similar to JSON, but with some differences.
+    /// See [the wiki](https://minecraft.gamepedia.com/NBT_format#SNBT_format) for more information.
+    /// # Example
+    /// ```
+    /// use nbt::Value;
+    /// use nbt::SNBTReader;
+    /// let value = SNBTReader::from_snbt("1f").unwrap();
+    /// assert_eq!(value, Value::Float(1.0));
+    /// ```
+    pub fn from_snbt(snbt: &str) -> Result<Value> {
+        SNBTReader::new(snbt).read()
+    }
+}
+
+pub struct SNBTWriter {
+    output: String,
+}
+
+impl SNBTWriter {
+    pub fn new() -> Self {
+        Self {
+            output: String::new(),
+        }
+    }
+
+    fn write_string(&mut self, s: &str) {
+        let mut need_quote = false;
+        for c in s.chars() {
+            if !matches!(c, 'a'..='z' | 'A'..='Z' | '0'..='9' | '_' | '-' | '+' | '.') {
+                need_quote = true;
+                break;
+            }
+        }
+        if need_quote {
+            self.output.push('"');
+            for c in s.chars() {
+                match c {
+                    '"' => self.output.push_str("\\\""),
+                    '\\' => self.output.push_str("\\\\"),
+                    _ => self.output.push(c),
+                }
+            }
+            self.output.push('"');
+        } else {
+            self.output.push_str(s);
+        }
+    }
+
+    fn write_primitive_array<'a>(
+        &mut self,
+        prefix: &str,
+        iter: impl Iterator<Item = &'a (impl Into<Value> + 'a + Copy)>,
+    ) {
+        self.output.push('[');
+        self.output.push_str(prefix);
+        let mut first = true;
+        for v in iter {
+            if !first {
+                self.output.push(',');
+            }
+            first = false;
+            self.write_element(&(*v).into());
+        }
+        self.output.push(']');
+    }
+
+    fn write_primitive(&mut self, postfix: &str, value: impl ToString) {
+        self.output.push_str(&value.to_string());
+        self.output.push_str(postfix);
+    }
+
+    fn write_list(&mut self, list: &List) {
+        macro_rules! variant_impl {
```

Comment:
It would be better to use a lambda here instead of a macro if possible.

Reply:
Help needed.

Reply:
Actually I'm alright with this.